gadgets/trace_ssl/program.bpf.c

```diff
@@ -0,0 +1,432 @@
+	gadget_timestamp timestamp;
+	u64 latency_ns;
+	u32 pid;
+	u32 tid;
+	u32 uid;
+	u32 gid;
+	u32 len;
+	u64 retval;
+	char comm[TASK_COMM_LEN];
+	u8 buf[MAX_BUF_SIZE];
+};
+
+#define BASE_EVENT_SIZE ((size_t)(&((struct event *)0)->buf))
+#define EVENT_SIZE(X) (BASE_EVENT_SIZE + ((size_t)(X)))
+
+GADGET_TRACER_MAP(events, 1024 * 256);
+GADGET_TRACER(ssl, events, event);
+
+const volatile bool record_data = true;
+GADGET_PARAM(record_data);
+
+/* used for context between uprobes and uretprobes of all operations */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32); // tid
+	__type(value, u64); // mntns_id
+} mntns_context SEC(".maps");
+
+/* used for context between uprobes and uretprobes of ssl operations */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32); // tid
+	__type(value, u64); // timestamp
+} ssl_starttime SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32); // tid
+	__type(value, void *); // buffer
+} ssl_bufs SEC(".maps");
+
+/* used for context between uprobes and uretprobes of libcrypto operations */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32); // tid
+	__type(value, u64); // timestamp
+} crypto_starttime SEC(".maps");
+
+/**
+ * clean up the maps when a thread terminates,
+ * because there may be residual data in the map
+ * if a userspace thread is killed between a uprobe and a uretprobe
+ */
+SEC("tracepoint/sched/sched_process_exit")
+int trace_sched_process_exit(void *ctx)
+{
+	u32 tid;
+
+	tid = (u32)bpf_get_current_pid_tgid();
+	bpf_map_delete_elem(&mntns_context, &tid);
+	bpf_map_delete_elem(&ssl_starttime, &tid);
+	bpf_map_delete_elem(&ssl_bufs, &tid);
+	bpf_map_delete_elem(&crypto_starttime, &tid);
+	return 0;
+}
+
+static __always_inline int probe_ssl_rw_enter(struct pt_regs *ctx, void *buf)
+{
+	u64 mntns_id;
+	u32 tid;
+	u64 ts;
+
+	mntns_id = gadget_get_mntns_id();
+	if (gadget_should_discard_mntns_id(mntns_id))
+		return 0;
+
+	tid = (u32)bpf_get_current_pid_tgid();
+	ts = bpf_ktime_get_boot_ns();
+
+	bpf_map_update_elem(&mntns_context, &tid, &mntns_id, BPF_ANY);
+	bpf_map_update_elem(&ssl_bufs, &tid, (u64 *)&buf, BPF_ANY);
+	bpf_map_update_elem(&ssl_starttime, &tid, &ts, BPF_ANY);
+	return 0;
+}
+
+static __always_inline int probe_ssl_rw_exit(struct pt_regs *ctx,
+					     enum operation op)
+{
+	u64 *mntns_ptr;
+	int len;
+	u64 *bufp;
+	u64 *tsp;
+	u32 buf_copy_size;
+	struct event *event;
+	u64 pid_tgid;
+	u64 uid_gid;
+	u32 pid;
+	u32 tid;
+	u64 ts;
+
+	pid_tgid = bpf_get_current_pid_tgid();
+	uid_gid = bpf_get_current_uid_gid();
+	pid = pid_tgid >> 32;
+	tid = (u32)pid_tgid;
+	ts = bpf_ktime_get_boot_ns();
+
+	len = PT_REGS_RC(ctx);
+	if (len <= 0) // no data
+		goto clean;
+
+	bufp = bpf_map_lookup_elem(&ssl_bufs, &tid);
+	if (bufp == 0)
+		goto clean;
+	bpf_map_delete_elem(&ssl_bufs, &tid);
```

Comment:
Why calling `bpf_map_delete_elem` here? It will be done at the end of the function anyway.